Halmap/Features/SongStorage/SongStorageViewModel.swift

```diff
@@ -0,0 +1,79 @@
+//
+//  SongStorageViewModel.swift
+//  Halmap
+//
+//  Created by JeonJimin on 2023/09/24.
+//
+
+import SwiftUI
+
+final class SongStorageViewModel: ObservableObject {
+    let dataManager: DataManager
+    let persistence: PersistenceController
+    let topEdge: CGFloat
+    let maxHeight: CGFloat = 216
+    
+    @Published var offset: CGFloat = 0
+    @Published var isDraged = false
+    
+    init(dataManager: DataManager, persistence: PersistenceController, topEdge: CGFloat) {
+        self.dataManager = dataManager
+        self.persistence = persistence
+        self.topEdge = topEdge
+    }
+    
+    func deleteSong(favoriteSong: CollectedSong) {
+        persistence.deleteSongs(song: favoriteSong)
+    }
+    
+    func makeSong(favoriteSong: CollectedSong) -> Song {
+        Song(
+            id: favoriteSong.safeId,
+            type: favoriteSong.type ,
+            title: favoriteSong.safeTitle ,
+            lyrics: favoriteSong.safeLyrics,
+            info: favoriteSong.safeInfo,
+            url: favoriteSong.safeUrl
+        )
+    }
+    private func makeSongInfo(favoriteSong: CollectedSong) -> SongInfo {
+        SongInfo(
+            id: favoriteSong.safeId,
+            team: favoriteSong.safeTeam,
+            type: favoriteSong.type ,
+            title: favoriteSong.safeTitle ,
+            lyrics: favoriteSong.safeLyrics,
+            info: favoriteSong.safeInfo,
+            url: favoriteSong.safeUrl
+        )
+    }
+    
+    func fetchSongImageName(favoriteSong: CollectedSong) -> String {
+        dataManager.checkSeasonSong(data: makeSongInfo(favoriteSong: favoriteSong)) ? "\(favoriteSong.team ?? "")23" : "\( favoriteSong.team ?? "NC")\(favoriteSong.type ? "Player" : "Album")"
+    }
+    
+    func handleOffsetChange(threshold: CGFloat) {
+        if self.offset > threshold {
+            withAnimation {
+                self.isDraged = false
+            }
+        } else if self.offset < 0 {
+            withAnimation {
+                self.isDraged = true
+            }
+        }
+    }
+    
+    func calculateOpacity(listCount: Int, isDefaultHeader: Bool, screenHeight: CGFloat) -> Double {
```

Comment:
해당 함수를 사용하는 모든 곳에서 screenHeight 값이 동일하던데, 확장을 고려하신거라면 default value를 UIScreen.screenHeight를 준다면, 사용하는 쪽에서 코드가 더 간결해질 것 같아요

 `viewModel.calculateOpacity(listCount: collectedSongs.count, isDefaultHeader: true)`

```suggestion
    func calculateOpacity(listCount: Int, isDefaultHeader: Bool, screenHeight: CGFloat = UIScreen.screenHeight) -> Double {
```

Reply:
확장성보다는 뷰모델 내부에서 UIScreen을 선언하여 사용하는 경우를 방지하기 위함이기때문에 기본값을 넣는 경우에도 마찬가지로 내부에서 선언하게되어 하지 않는기 좋을 것 같습니다

Reply:
아 UI screen이 뷰에 대한 로직이기때문이라는거죠?
그럼 맨 위에 리뷰도 닫겠습니다!

검색 뷰도 UI 로직 분리하겠습니다 !